### PR TITLE
Begin adding unit tests for test-bot

### PR DESCRIPTION
**Pull request by tdsmith**
I'd like to start adding some unit tests for test-bot to allow us to refactor parts of it with more confidence. I'm envisioning a series of small PRs. Ultimately I'd like test-bot to become more testable.

### CODE REVIEW

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+class TestbotCommandTests < Homebrew::TestCase
+  def with_environment(h)
+    old = Hash[h.keys.map { |k| [k, ENV[k]] }]
+    ENV.update h
+    begin
+      yield
+    ensure
+      ENV.update old
+    end
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: You might want to use `ENV.to_hash` and `ENV.replace` for backing up and restoring the environment. This would also improve consistency with other tests, e.g. [these lines from `test_utils.rb`](https://github.com/Homebrew/brew/blob/cab97cf4d782ce36762f0cd615ca59718f55b0d5/Library/Homebrew/test/test_utils.rb#L36-L44), and I believe also other environment handling code we have in Homebrew.

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+require "testing_env"
+require "dev-cmd/test-bot"
+
+class TestbotCommandTests < Homebrew::TestCase
+  def with_environment(h)
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: Please use a more self-descriptive argument name like `extra_env`. (Feel free to come up with an even better name!)

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    end
+  end
+
+  def test_resolve_test_tap
+    tap = Homebrew::resolve_test_tap
+    assert_equal tap, nil, "Should return nil if no tap slug provided"
+
+    slug = "spam/homebrew-eggs"
+    url = "https://github.com/#{slug}.git"
+    pairs = [
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: `pairs` is a nice name for the structure of this thing, but what does it actually contain? (Subtle hint to maybe come up with a better variable name.)

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      {"TRAVIS_REPO_SLUG" => slug},
+      {"UPSTREAM_BOT_PARAMS" => "--tap=#{slug}"},
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: That `--tap=spam/homebrew-eggs` is recognized is more of a convenience (no need to adjust the Travis CI repository slug beforehand), but I'd actually expect this to test `--tap=spam/eggs` or possibly both.

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      {"UPSTREAM_GIT_URL" => url},
+      {"GIT_URL" => url},
+    ]
+
+    predicate = Proc.new do
+      tap = Homebrew::resolve_test_tap
+      assert_kind_of Tap, tap
+      assert_equal tap.user, "spam"
+      assert_equal tap.repo, "eggs"
+    end
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: This is a nice way to avoid duplicate code. Does it also provide a meaningful output if any of the assertions fail, i.e. is it easy to understand for which case and which `assert_*` invocation the failure occurred?

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    ENV.update h
+    begin
+      yield
+    ensure
+      ENV.update old
+    end
+  end
+
+  def test_resolve_test_tap
+    tap = Homebrew::resolve_test_tap
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: Prefer `Homebrew.resolve_test_tap` for invoking methods (should have been pointed out by `brew style`).

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    assert_equal tap, nil, "Should return nil if no tap slug provided"
```
Comment (UniqMartin) [inline Library/Homebrew/test/test_cmd_testbot.rb]: Use `assert_nil`.

**Comment by UniqMartin** [comment]
Thanks for starting this! I allowed myself to leave some code comments with questions, suggestions, and the usual amount of style nit picking. 🙈

**Comment by tdsmith** [comment]
I noticed. :D

File: Library/Homebrew/test/test_cmd_testbot.rb (hunk 1 above)
Comment (tdsmith) [inline Library/Homebrew/test/test_cmd_testbot.rb]: Nice, thanks!

File: Library/Homebrew/test/test_cmd_testbot.rb (hunk 5 above)
Comment (tdsmith) [inline Library/Homebrew/test/test_cmd_testbot.rb]: It provides good feedback about which assertion but not which case; I'll add a message parameter to the block.

**Comment by tdsmith** [comment]
Updated to address your feedback; thanks for taking a look.

**Comment by UniqMartin** [comment]
Great, including the style changes that I haven't pointed out because I don't have them memorized! (But I don't have to, because I have my editor set up to highlight RuboCop violations during editing. If only `brew audit` would always output line and—if applicable—column information, it could be used similarly.)

**Comment by MikeMcQuaid** [comment]
> I'd like to start adding some unit tests for test-bot to allow us to refactor parts of it with more confidence. I'm envisioning a series of small PRs. Ultimately I'd like test-bot to become more testable.

Sorry for the delayed response here, vacation etc. I guess I'm trying to figure out here: what's the motivation for end-users? I'm a pretty hard 👎 on unit testing stuff like the test-bot as-is; I think it makes refactoring more rather than less difficult. I think we could exercise it better with a single, high (or full) coverage integration test which installs multiple no-op formulae which have dependency relationships. As-is my concern is we're just testing that the internal implementation behaves as expected which makes the internal implementation harder to change to a better approach.

File: Library/Homebrew/test/test_cmd_testbot.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+    environments.each do |pair|
+      with_environment(pair) do
+        predicate.call pair.to_s
+      end
+    end
+
+    ARGV.expects(:value).with("tap").returns(slug)
+    predicate.call "ARGV"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/test_cmd_testbot.rb]: It's pretty hard to follow on what's going on here, can you elaborate?

Comment (tdsmith) [inline Library/Homebrew/test/test_cmd_testbot.rb]: `resolve_test_tap` is responsive to parameters delivered through either the environment or through ARGV. This tests delivering parameters through each of the supported environments and then ARGV.

**Comment by tdsmith** [comment]
I think I agree that these unit tests do not solve the right problem, but writing them was a useful exercise for me! My biggest concern about test-bot is that several of the methods are very long and are decomposable into smaller units which can be individually tested for correctness. Another is that argument passing has a lot of dependencies on global state. I think it makes this harder to reason about and fix test-bot behaviors. I agree that any refactoring is going to need to depend on good integration tests.
